lualib/resty/kong/tls.lua

```diff
@@ -131,4 +154,38 @@ do
 end
 
 
+do
+    local ALLOWED_PHASES = {
+        ['rewrite'] = true,
+        ['balancer'] = true,
+        ['access'] = true,
+    }
+
+    function _M.set_upstream_cert_and_key(chain, key)
+        if not ALLOWED_PHASES[get_phase()] then
+            error("API disabled in the current context", 2)
+        end
+
+        if not chain or not key then
+            error("chain and key must not be nil")
```

Comment:
This will be triggered when either `chain` or `key` are `false` too. That could be fixed by changing the error message to `chain and key must be present`.

However what happens if `chain` or `key` is something more ... exotic, like a table value? Someone could do `:set_upstream_cert_and_key` instead of `.set_upstream_cert_and_key` for example.

Is `C.ngx_http_lua_kong_ffi_set_upstream_client_cert_and_key` able to handle such errors? If not, I suggest changing this to a more explicit check for types.

What about empty strings? Are they considered valid keys?

```
if key == "" or type(key) ~= "string" then
  error("key must be a non-empty string")
end

if type(chain) ~= ... then -- not sure about this one, don't know what STACK_OF(X509) means
  ...
end
```

Reply:
In `resty.*` module,s arguments validation is typically kept to a minimum. We could have stricter arguments validation in our `kong.*` PDK module though.

Reply:
And PDK does https://github.com/Kong/kong/pull/4797/files#diff-0a162c8833b46ddf83218f6afa5a32cdR452! Although we are limited about what kind of validation we can do. If the user passed the wrong type of `cdata` pointer in we have no way to tell. But that's the same situation with `ngx.ssl` right now as well.